### PR TITLE
[SM-113] Delete Projects Dialog

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

The purpose of this PR is to implement the delete projects dialog for the secrets manager project.

## Code changes

- apps/web/src/locales/en/messages.json

Updating EmptyList message to NoItems messages to match with the NoItemsComponent. 
Adding new messages needed for the delete projects dialog. 

- bitwarden_license/bit-web/src/app/sm/layout/dialogs/bulk-status-dialog.component.html
bitwarden_license/bit-web/src/app/sm/layout/dialogs/bulk-status-dialog.component.ts

A dialog component used to display error messages for bulk operations.

- bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html
bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.ts

Code for the delete projects dialog component. Handles single and bulk deletes.

- bitwarden_license/bit-web/src/app/sm/projects/project.service.ts

Adding in code to handle deletes in the project service.

- bitwarden_license/bit-web/src/app/sm/projects/projects-list.component.html
bitwarden_license/bit-web/src/app/sm/projects/projects-list.component.ts

Updating events for delete and adding in the NoItems display to handle empty project list. 

- bitwarden_license/bit-web/src/app/sm/projects/projects.component.ts
bitwarden_license/bit-web/src/app/sm/projects/projects.component.html

Updating to handle project delete events

- bitwarden_license/bit-web/src/app/sm/secrets/secrets-list.component.html bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts-list.component.html

Updating i18n text to the new NoItems format.

- bitwarden_license/bit-web/src/app/sm/shared/sm-shared.module.ts

Adding BulkStatusDialogComponent to the module.

###  No Items Project

image

### Single Delete

image

image

image

image

image

### Bulk Delete

image

image

image

image

image

### Singe Delete Error

image

### Bulk Delete Error

image

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        columnTitle: "projectName",
+        message: "bulkDeleteProjectsErrorMessage",
+        details: bulkStatusResults,
+      },
+    });
+  }
+
+  private get dialogConfirmationMessage() {
+    return this.data.projects.length === 1
+      ? this.i18nService.t("deleteProjectConfirmMessage", this.data.projects[0].name)
```
Comment (coltonhurst) [inline bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.ts:101]: What if `this.data.projects` has no elements?

File: bitwarden_license/bit-web/src/app/sm/layout/dialogs/bulk-status-dialog.component.html [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        <tr bitRow *ngFor="let detail of data.details">
+          <td bitCell>{{ detail.name }}</td>
+          <td bitCell>{{ detail.errorMessage }}</td>
+        </tr>
+      </ng-container>
+    </bit-table>
+  </div>
+
+  <div bitDialogFooter class="tw-flex tw-gap-2">
+    <button bitButton buttonType="primary" bitDialogClose (click)="dialogRef.close()">
```
Comment (coltonhurst) [inline bitwarden_license/bit-web/src/app/sm/layout/dialogs/bulk-status-dialog.component.html:26]: Based on Oscar's recent comments I think we should be adding `type=button` to any close / cancel buttons

File: bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        <bit-label>{{ dialogConfirmationLabel }}</bit-label>
+        <input bitInput formControlName="confirmDelete" />
+      </bit-form-field>
+    </div>
+
+    <div bitDialogFooter class="tw-flex tw-gap-2">
+      <button bitButton buttonType="danger" [loading]="form.loading">
+        {{ title | i18n }}
+      </button>
+      <button bitButton buttonType="secondary" bitDialogClose (click)="dialogRef.close()">
```
Comment (coltonhurst) [inline bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html:33]: `type=button` here also, see similar comment for explanation

File: bitwarden_license/bit-web/src/app/sm/layout/dialogs/bulk-status-dialog.component.html (hunk 2 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/layout/dialogs/bulk-status-dialog.component.html]: Resolved in https://github.com/bitwarden/clients/pull/3777/commits/727d62c6443cbdf4d5d11262f6c89b81581da2a9

File: bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html (hunk 3 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html]: Resolved in https://github.com/bitwarden/clients/pull/3777/commits/727d62c6443cbdf4d5d11262f6c89b81581da2a9

File: bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.ts (hunk 1 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.ts:110]: Good catch. 

I added an angular OnInit method for each dialog and validated the injected data there. 

If something calls the dialog and passes bad data, the dialog will automatically close and throw an error.

I'm open to other solutions, but this was the only thing I could think of doing.

Comment (coltonhurst) [reply bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.ts:110]: Thanks for this!

I spoke with @Hinton about it - sometime in the future (possibly next year?) we will be turning on TypeScript strict typing. Specifically, `strictNullChecks` set as `true` should eliminate the need for these checks. For now though, having checks in `ngOnInit()` is the way to go.

One last request - Oscar mentioned we should add "todo" comments so we can remove the checks once the strict typing is turned on. Do you mind adding that?

Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.ts:110]: Comment added in https://github.com/bitwarden/clients/pull/3777/commits/9deb74844c201548583cbc6091189b3baf23be79

**Review by coltonhurst — APPROVED** [review]

File: bitwarden_license/bit-web/src/app/sm/layout/dialogs/bulk-status-dialog.component.html [hunk 4]
```diff
@@ -0,0 +1,30 @@
+<bit-dialog dialogSize="default">
+  <ng-container bitDialogTitle>{{ data.title | i18n }}</ng-container>
+  <span bitDialogTitle class="tw-text-sm tw-normal-case tw-text-muted">
+    {{ " " + data.details.length + " " }} {{ data.subTitle | i18n }}
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/sm/layout/dialogs/bulk-status-dialog.component.html:4]: Do we need the spacing here?

File: bitwarden_license/bit-web/src/app/sm/layout/dialogs/bulk-status-dialog.component.html [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        <tr bitRow *ngFor="let detail of data.details">
+          <td bitCell>{{ detail.name }}</td>
+          <td bitCell>{{ detail.errorMessage }}</td>
+        </tr>
+      </ng-container>
+    </bit-table>
+  </div>
+
+  <div bitDialogFooter class="tw-flex tw-gap-2">
+    <button bitButton buttonType="primary" bitDialogClose (click)="dialogRef.close()" type="button">
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/sm/layout/dialogs/bulk-status-dialog.component.html:26]: We already use the `bitDialogClose`, which means we shouldn't need to also have a click handler.

File: bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html [hunk 6]
```diff
@@ -0,0 +1,44 @@
+<form #form [formGroup]="formGroup" (ngSubmit)="submit()" [appApiAction]="formPromise">
+  <bit-dialog dialogSize="small">
+    <ng-container bitDialogTitle>{{ title | i18n }}</ng-container>
+    <span
+      bitDialogTitle
+      class="tw-text-sm tw-normal-case tw-text-muted"
+      *ngIf="data.projects.length == 1"
+    >
+      {{ " " + data.projects[0].name }}
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html:9]: Same here, do we actually need the spacing? Or is it just increasing the regular distance in which case do we want to use margin instead?

File: bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    </span>
+    <span
+      bitDialogTitle
+      class="tw-text-sm tw-normal-case tw-text-muted"
+      *ngIf="data.projects.length > 1"
+    >
+      {{ " " + data.projects.length + " " }} {{ "projects" | i18n }}
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html:16]: Spacing?

File: bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+    <div bitDialogFooter class="tw-flex tw-gap-2">
+      <button bitButton buttonType="danger" [loading]="form.loading">
+        {{ title | i18n }}
+      </button>
+      <button
+        bitButton
+        buttonType="secondary"
+        bitDialogClose
+        (click)="dialogRef.close()"
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html:37]: I don't believe we need the click handler.

```suggestion
```

**Review by Hinton — COMMENTED** [review]
Some open questions.

File: bitwarden_license/bit-web/src/app/sm/layout/dialogs/bulk-status-dialog.component.html (hunk 4 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/layout/dialogs/bulk-status-dialog.component.html]: We do need spacing. I swapped to using `tw-gap-4` on a div to implement that spacing. 

 What do you think? 
 
https://github.com/bitwarden/clients/pull/3777/commits/2d15a15e195e68e757798cc04c8355299c4f4859

File: bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html (hunk 6 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html]: We do need spacing between the words. I swapped to using `tw-gap-4` on a div to implement that spacing. 

 What do you think? 
 
https://github.com/bitwarden/clients/pull/3777/commits/2d15a15e195e68e757798cc04c8355299c4f4859

File: bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html (hunk 7 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html]: We do need spacing between the words. I swapped to using `tw-gap-4` on a div to implement that spacing. 

 What do you think? 
 
https://github.com/bitwarden/clients/pull/3777/commits/2d15a15e195e68e757798cc04c8355299c4f4859

File: bitwarden_license/bit-web/src/app/sm/layout/dialogs/bulk-status-dialog.component.html (hunk 5 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/layout/dialogs/bulk-status-dialog.component.html:31]: When I remove (click)="dialogRef.close()" the button no longer closes out the dialog in Firefox or Chrome.

Am I missing something?

File: bitwarden_license/bit-web/src/app/sm/layout/dialogs/bulk-status-dialog.component.html (hunk 4 above)
Comment (Hinton) [reply bitwarden_license/bit-web/src/app/sm/layout/dialogs/bulk-status-dialog.component.html]: The class doesn't do anything on my machine. (Chrome)

File: bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+      <span class="tw-text-sm tw-normal-case tw-text-muted" *ngIf="data.projects.length == 1">
+        {{ data.projects[0].name }}
+      </span>
+      <span class="tw-text-sm tw-normal-case tw-text-muted" *ngIf="data.projects.length > 1">
+        {{ data.projects.length }}
+      </span>
+      <span class="tw-text-sm tw-normal-case tw-text-muted" *ngIf="data.projects.length > 1">
+        {{ "projects" | i18n }}
+      </span>
+    </div>
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html:14]:
```suggestion
    <ng-container bitDialogTitle>
      <span>{{ title | i18n }}</span>
      <span class="tw-text-sm tw-normal-case tw-text-muted">
        <ng-container *ngIf="data.projects.length == 1">
          {{ data.projects[0].name }}
        </ng-container>
        <ng-container *ngIf="data.projects.length > 1">
          {{ data.projects.length }}
          {{ "projects" | i18n }}
        </ng-container>
      </span>
    </ng-container>
```

Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html]: Resolved in https://github.com/bitwarden/clients/pull/3777/commits/0cad733924d348e72b51bbae6459a9c947d550e3

File: bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html (hunk 7 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html]: Resolved in https://github.com/bitwarden/clients/pull/3777/commits/0cad733924d348e72b51bbae6459a9c947d550e3

File: bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html (hunk 6 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/projects/dialog/project-delete-dialog.component.html]: Resolved in https://github.com/bitwarden/clients/pull/3777/commits/0cad733924d348e72b51bbae6459a9c947d550e3

File: bitwarden_license/bit-web/src/app/sm/projects/project.service.ts [hunk 10]
```diff
@@ -63,6 +65,22 @@ export class ProjectService {
...
+    const r = await this.apiService.send("POST", "/projects/delete", projectIds, true, true);
+
+    const bulkResponse: BulkOperationStatus[] = [];
+    r.data.forEach((element: { id: string; error: string }) => {
+      const bulkOperationStatus = new BulkOperationStatus();
+      bulkOperationStatus.id = element.id;
+      bulkOperationStatus.name = projects.find((project) => project.id == element.id).name;
+      bulkOperationStatus.errorMessage = element.error;
+      bulkResponse.push(bulkOperationStatus);
+    });
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/sm/projects/project.service.ts:79]: Are we better of doing a map here?

**Review by Hinton — APPROVED** [review]

File: bitwarden_license/bit-web/src/app/sm/projects/project.service.ts (hunk 10 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/projects/project.service.ts]: 100% resolved in https://github.com/bitwarden/clients/pull/3777/commits/4a20a5f2af72165c6865c86ac51e77c7e8104308

**Review by Hinton — APPROVED** [review]

**Pull request merged**